stan/math/prim/mat/meta/is_eigen.hpp

```diff
@@ -36,5 +36,28 @@ struct is_eigen<
     T, std::enable_if_t<internal::is_eigen_base<std::decay_t<T>>::value>>
     : std::true_type {};
 
+namespace internal {
+template <typename... Ts>
+struct make_void {
+  typedef void type;
+};
+template <typename... Ts>
+using void_t = typename make_void<Ts...>::type;
+
+template <class, class = void>
+struct is_eigen_matrix_impl : std::false_type {};
+template <class T>
+struct is_eigen_matrix_impl<T, void_t<typename T::Matrix()>> : std::true_type {
+};
+
+template <class T>
+struct is_eigen_matrix_impl<T, void_t<typename T::SparseMatrix()>>
+    : std::true_type {};
+
+}  // namespace internal
+
+template <typename T>
+struct is_eigen_matrix : internal::is_eigen_matrix_impl<std::decay_t<T>> {};
```

Comment:
@SteveBronder  The tests for this fail with clang++ but pass with g++:

```
test/unit/math/prim/mat/meta/is_eigen_test.cpp:41: Failure
Value of: (is_eigen_matrix<Eigen::Matrix<double, -1, -1>>::value)
  Actual: false
Expected: true
test/unit/math/prim/mat/meta/is_eigen_test.cpp:42: Failure
Value of: (is_eigen_matrix<Eigen::SparseMatrix<double>>::value)
  Actual: false
Expected: true
test/unit/math/prim/mat/meta/is_eigen_test.cpp:45: Failure
Value of: (is_eigen_matrix<const Eigen::Matrix<double, -1, -1>>::value)
  Actual: false
Expected: true
test/unit/math/prim/mat/meta/is_eigen_test.cpp:46: Failure
Value of: (is_eigen_matrix<Eigen::SparseMatrix<double>&>::value)
  Actual: false
Expected: true
```
Any idea why? Should we split this to a separate PR? Your call.

Reply:
Aaaah yes I saw that problem on clang as well. Let me look tomorrow I bet there's an easy fix